### DISCUSSION TITLE
How to get old value or time derivative of a defined ADMaterial property?

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#26475] [Q&A General] [answered]
Hi all,

I would like to calculate the time derivative of a defined ADMaterial property. But I can not use `adCoupledDot` like coupledVar. Then, I try to calculate the time derivative use: (material_current - material_old)/_dt. But I can not get old value of ADMaterial, since `getADMaterialPropertyOld` does not exist. So, what should I do? Using `getMaterialPropertyOld`? But it is for non-AD material.

Thanks.

**Comment by lindsayad** [comment, marked as answer]
Yes, you can use `getMaterialPropertyOld` for both AD and non-AD material properties since the old state is guaranteed to not depend on the nonlinear system degrees of freedom

**Reply by SomnusYu** [reply]
Thanks!

**Reply by SomnusYu** [reply]
Hi,

I used the `getMaterialPropertyOld` to get the value of `material_old` and then calculated as: `(material_current - material_old)/_dt`. 
But some warnings occurred:
![image](https://github.com/idaholab/moose/assets/118275988/e7647e49-8cc5-44dd-b191-34dbf9b0fef5)

Do these warnings matter? Can I ignore these warnings?

**Reply by GiudGiud** [reply]
initQpStatefulProperties is used to define the inital value of the property. So that's the very first value of material_old for example. 

You can ignore them if the initial value is 0. But it s not best practice.

**Reply by SomnusYu** [reply]
ok
